drivers/spi/spi_ite_it8xxx2.c

```diff
@@ -0,0 +1,496 @@
+		err = -ETIMEDOUT;
+	}
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	return err;
+
+}
+
+static int it8xxx2_transceive(const struct device *dev,
+			    const struct spi_config *config,
+			    const struct spi_buf_set *tx_bufs,
+			    const struct spi_buf_set *rx_bufs)
+{
+	size_t nbrx = 0;
+	size_t nbtx = 0;
+
+	const struct spi_buf *ptx = tx_bufs->buffers;
+	const struct spi_buf *prx = rx_bufs->buffers;
+
+	int err = 0;
+	int ret = 0;
+
+	ret = k_mutex_lock(&it8xxx2_mutex, K_FOREVER);
+
+	if (tx_bufs != NULL) {
+		nbtx = tx_bufs->count;
+	}
+
+	while (nbtx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	if (rx_bufs != NULL) {
+		nbrx = rx_bufs->count;
+	}
+
+	while (nbrx--) {
+		err = spi_it8xxx2_trans(dev, config, ptx, prx);
+		if (err != 0) {
+			LOG_ERR("[kern][spi_it8xxx2_trans] Returned value is not 0\r\n");
+		}
+	}
+
+	if (ret == -EAGAIN) {
+		LOG_ERR("%s: Timeout", __func__);
+		err = ret;
+	}
+
+	k_mutex_unlock(&it8xxx2_mutex);
+
+	return err;
+
+}
+
+static int it8xxx2_transceive_sync(const struct device *dev,
+				 const struct spi_config *config,
+				 const struct spi_buf_set *tx_bufs,
+				 const struct spi_buf_set *rx_bufs)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+
+#ifdef CONFIG_SPI_ASYNC
+static int it8xxx2_transceive_async(const struct device *dev,
+				  const struct spi_config *config,
+				  const struct spi_buf_set *tx_bufs,
+				  const struct spi_buf_set *rx_bufs,
+				  struct k_poll_signal *async)
+{
+	return it8xxx2_transceive(dev, config, tx_bufs, rx_bufs);
+}
+#endif
+
+static int it8xxx2_release(const struct device *dev,
+			 const struct spi_config *config)
+{
+	k_sem_give(&it8xxx2_sem);
+	return 0;
+}
+
+static void spi_it8xxx2_isr(const struct device *dev)
+{
+	const struct spi_it8xxx2_config *cfg = dev->config;
+	struct spi_it8xxx2_regs *const spi_regs =
+		(struct spi_it8xxx2_regs *)cfg->base_addr;
+
+	irq_disable(cfg->irq_no);
+
+	spi_regs->INTSTS |= SPICMDQEND;
+	ite_intc_isr_clear(cfg->irq_no);
+	spi_regs->SPICTRL5 &= ~(CH1SELCMDQ | CH0SELCMDQ);
+	k_sem_give(&it8xxx2_sem);
+	chip_permit_idle();
+}
+
+static int spi_it8xxx2_init(const struct device *dev)
+{
+	const struct spi_it8xxx2_config *cfg = dev->config;
+	int status;
+
+	status = pinctrl_apply_state(cfg->pcfg, PINCTRL_STATE_DEFAULT);
+
+	if (status < 0) {
+		LOG_ERR("Failed to configure SPI pins\r\n");
+		return status;
+	}
+
+	/* Shared IRQ */
+	if (!(cfg->inst_no)) {
+		/* Initialize Mutex */
+		k_mutex_init(&it8xxx2_mutex);
+		/* Initialize Semaphore */
+		k_sem_init(&it8xxx2_sem, 0, 1);
+		IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority),
+		spi_it8xxx2_isr, DEVICE_DT_INST_GET(0), 0);
+
+		it8xxx2_spi_cmdq_init(dev);
+		it8xxx2_spi_configure_sckfreq(dev);
+		it8xxx2_spi_configure_spimode(dev);
+		it8xxx2_spi_int_init(dev);
```

Comment:
I think you need to call these functions for every SPI instance used, correct?

Reply:
Dear Keith,

Well, theoretically, yes. However, the SPI channel 0 & 1 shares these configurations. I think that it'd be the limitation for the current IT8XXX2 chip.

Thanks again for guidance, sir.

Best regards,
BJ

Reply:
This has been modified to be invoked for each instance but so far the IT8XXX2 would not have the spi1.